services/src/layers/external.rs

```diff
@@ -0,0 +1,73 @@
+use async_trait::async_trait;
+use geoengine_datatypes::dataset::{DatasetId, LayerProviderId};
+use geoengine_datatypes::primitives::{RasterQueryRectangle, VectorQueryRectangle};
+use geoengine_operators::engine::{
+    MetaDataProvider, RasterResultDescriptor, VectorResultDescriptor,
+};
+use geoengine_operators::mock::MockDatasetDataSourceLoadingInfo;
+use geoengine_operators::source::{GdalLoadingInfo, OgrSourceDataset};
+
+use crate::datasets::listing::ProvenanceOutput;
+use crate::error::Result;
+
+use super::listing::LayerCollectionProvider;
+
+#[typetag::serde(tag = "type")]
+#[async_trait]
+pub trait ExternalLayerProviderDefinition:
+    CloneableLayerProviderDefinition + Send + Sync + std::fmt::Debug
+{
+    /// create the actual provider for data listing and access
+    async fn initialize(self: Box<Self>) -> Result<Box<dyn ExternalLayerProvider>>;
+
+    /// the type of the provider
+    fn type_name(&self) -> String;
+
+    /// name of the external data source
+    fn name(&self) -> String;
+
+    /// id of the provider
+    fn id(&self) -> LayerProviderId;
+}
+
+pub trait CloneableLayerProviderDefinition {
+    fn clone_boxed_provider(&self) -> Box<dyn ExternalLayerProviderDefinition>;
+}
+
+impl<T> CloneableLayerProviderDefinition for T
+where
+    T: 'static + ExternalLayerProviderDefinition + Clone,
+{
+    fn clone_boxed_provider(&self) -> Box<dyn ExternalLayerProviderDefinition> {
+        Box::new(self.clone())
+    }
+}
+
+impl Clone for Box<dyn ExternalLayerProviderDefinition> {
+    fn clone(&self) -> Box<dyn ExternalLayerProviderDefinition> {
+        self.clone_boxed_provider()
+    }
+}
+
+/// A provider of layers that are not hosted by Geo Engine itself but some external party
+// TODO: Authorization: the provider needs to accept credentials for the external data source.
```

Comment:
how will we handover credentials to the provider? i guess they need to be passed at some point?

Reply:
I think we should implement it in such a way, that the Context injects the user credentials into the Db (or a wrapper) s.th. we don't have to explicitly pass the session into the functions every time, because the Db already knows it. This makes it much easier as we don't need a generic for the session in the traits. However, IMHO, this should be done in a different task where we do that with all the Dbs.